dns.go

```diff
@@ -20,29 +20,32 @@ import (
 // OpportunisticResolver opportunistically tries encrypted DNS over TLS
 // using the local resolver.
 var OpportunisticResolver = &net.Resolver{
-	Dial:     opportunisticDial,
+	Dial:     getAnOpportunisticDialerFromDialFunc(nil),
 	PreferGo: true,
 }
 
-func opportunisticDial(ctx context.Context, network, address string) (net.Conn, error) {
-	host, port, _ := net.SplitHostPort(address)
-	if (port == "53" || port == "domain") && notBadServer(address) {
-		deadline, ok := ctx.Deadline()
-		if ok && deadline.After(time.Now().Add(2*time.Second)) {
-			var d net.Dialer
-			d.Timeout = time.Second
-			tlsAddr := net.JoinHostPort(host, "853")
-			tlsConf := tls.Config{InsecureSkipVerify: true}
-			conn, _ := tls.DialWithDialer(&d, "tcp", tlsAddr, &tlsConf)
-			if conn != nil {
-				return conn, nil
+func getAnOpportunisticDialerFromDialFunc(dial DialFunc) DialFunc {
+	if dial == nil {
+		var d net.Dialer
+		dial = d.DialContext
+	}
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		host, port, _ := net.SplitHostPort(address)
+		if (port == "53" || port == "domain") && notBadServer(address) {
+			deadline, ok := ctx.Deadline()
+			if ok && deadline.After(time.Now().Add(2*time.Second)) {
+				tlsAddr := net.JoinHostPort(host, "853")
+				rawConn, _ := dial(ctx, "tcp", tlsAddr)
```

Comment:
This is semantically very different from above.

My version ensures we have _at least_ 2 seconds before the deadline, and reserves one second to try and connect over TLS. If that fails, there's another second to try UDP/TCP.

Your version ensures we have the 2 seconds, but then possibly exhausts the full deadline (which is typically 5 seconds, I think), before trying the fallback. By then, the `ctx` is done and there's nothing else to do.

Can we even complete a TLS handshake over Tor quickly enough that there's still time for the fallback?

Reply:
I don't know what would be the right way to do it.

The way I would fix that is : create a cancel struct{} chan, a child cancelable ctx of the main ctx, and a 1 second ticker.
 Then fire a goroutine selecting over the cancel chan (if hit return) and the ticker (if hit cancel the ctx and return) then in the main goroutine defer a close of the cancel chan.
(in the spawned goroutine it would also be needed to stop the ticker with defer)

But that spawns a goroutine (and I don't really like that).

Reply:
I don't understand why a `context.WithTimeout` doesn't work.

You're creating a child context for the `DialContext`. Yes, that gets canceled after a second, but canceling the child does not cancel the parent. This is exactly how you're supposed to do subtask that don't starve your context. You can still do the fall back using the parent context.

Besides, if I'm not mistaken, you've just described how `context.WithTimeout` is implemented. 😀